### PR TITLE
fix: when marketsCpy slice shrinks, the onTick loop goes on

### LINKED ISSUE #8596: [Bug]: Consensus failure trying to leave opening auction for child when multiple markets try to use the same parent
### Problem encountered

Attempting to leave opening auction with a child market when there are two other markets claiming the same parent results in a consensus failure:
```
2023-06-26T18:26:21.225+0100    ERROR   tendermint      consensus/state.go:732  CONSENSUS FAILURE!!!    {"err": "runtime error: invalid memory address or nil pointer dereference", "stack": "goroutine 285 [running]:\nruntime/debug.Stack()
/usr/local/go/src/runtime/debug/stack.go:24 +0x65\ngithub.com/tendermint/tendermint/consensus.(*State).receiveRoutine.func2()
/home/jake-vega/go/pkg/mod/github.com/vegaprotocol/cometbft@v0.34.28-0.20230322133204-3d8588de736e/consensus/state.go:732 +0x4c\npanic({0x2f3baa0, 0x59a8c40})
/usr/local/go/src/runtime/panic.go:884 +0x212\ncode.vegaprotocol.io/vega/core/execution/future.(*Market).GetID(...)
/workspace/vega/core/execution/future/market.go:622\ncode.vegaprotocol.io/vega/core/execution.(*Engine).OnTick(0xc0003c8000, {0x3db7460, 0xc005436150}, {0x1?, 0x3dd5420?, 0x0?})
/workspace/vega/core/execution/engine.go:897 +0x2fc\ncode.vegaprotocol.io/vega/core/vegatime.(*Svc).notify(...)
/workspace/vega/core/vegatime/service.go:104\ncode.vegaprotocol.io/vega/core/vegatime.(*Svc).SetTimeNow(0xc0012cc500, {0x3db7460, 0xc005436150}, {0x40?, 0x2ddfb20?, 0x0?})
/workspace/vega/core/vegatime/service.go:77 +0x24f\ncode.vegaprotocol.io/vega/core/processor.(*App).OnBeginBlock(_, {{0xc004709c80, 0x20, 0x20}, {{0xb, 0x1}, {0xc0079449a0, 0xb}, 0x15a, {0x17a77224, ...}, ...}, ...})
/workspace/vega/core/processor/abci.go:797 +0x747\ncode.vegaprotocol.io/vega/core/blockchain/abci.(*App).BeginBlock(_, {{0xc004709c80, 0x20, 0x20}, {{0xb, 0x1}, {0xc0079449a0, 0xb}, 0x15a, {0x17a77224, ...}, ...}, ...})
/workspace/vega/core/blockchain/abci/abci.go:46 +0x75\ncode.vegaprotocol.io/vega/cmd/vega/commands/node.(*appW).BeginBlock(_, {{0xc004709c80, 0x20, 0x20}, {{0xb, 0x1}, {0xc0079449a0, 0xb}, 0x15a, {0x17a77224, ...}, ...}, ...})
/workspace/vega/cmd/vega/commands/node/app_wrapper.go:64 +0x55\ngithub.com/tendermint/tendermint/abci/client.(*localClient).BeginBlockSync(_, {{0xc004709c80, 0x20, 0x20}, {{0xb, 0x1}, {0xc0079449a0, 0xb}, 0x15a, {0x17a77224, ...}, ...}, ...})
/home/jake-vega/go/pkg/mod/github.com/vegaprotocol/cometbft@v0.34.28-0.20230322133204-3d8588de736e/abci/client/local_client.go:280 +0x118\ngithub.com/tendermint/tendermint/proxy.(*appConnConsensus).BeginBlockSync(_, {{0xc004709c80, 0x20, 0x20}, {{0xb, 0x1}, {0xc0079449a0, 0xb}, 0x15a, {0x17a77224, ...}, ...}, ...})
/home/jake-vega/go/pkg/mod/github.com/vegaprotocol/cometbft@v0.34.28-0.20230322133204-3d8588de736e/proxy/app_conn.go:81 +0x55\ngithub.com/tendermint/tendermint/state.execBlockOnProxyApp({0x3db59b0?, 0xc00029ccc8}, {0x3dcbf80, 0xc001517a40}, 0xc003f825a0, {0x3dde6e8, 0xc0005be390}, 0x159?)
/home/jake-vega/go/pkg/mod/github.com/vegaprotocol/cometbft@v0.34.28-0.20230322133204-3d8588de736e/state/execution.go:307 +0x3dd\ngithub.com/tendermint/tendermint/state.(*BlockExecutor).ApplyBlock(_, {{{0xb, 0x1}, {0xc002c82940, 0x7}}, {0xc002c82950, 0xb}, 0x1, 0x159, {{0xc0059b2240, ...}, ...}, ...}, ...)
/home/jake-vega/go/pkg/mod/github.com/vegaprotocol/cometbft@v0.34.28-0.20230322133204-3d8588de736e/state/execution.go:140 +0x171\ngithub.com/tendermint/tendermint/consensus.(*State).finalizeCommit(0xc002c92380, 0x15a)
/home/jake-vega/go/pkg/mod/github.com/vegaprotocol/cometbft@v0.34.28-0.20230322133204-3d8588de736e/consensus/state.go:1661 +0xafd\ngithub.com/tendermint/tendermint/consensus.(*State).tryFinalizeCommit(0xc002c92380, 0x15a)
/home/jake-vega/go/pkg/mod/github.com/vegaprotocol/cometbft@v0.34.28-0.20230322133204-3d8588de736e/consensus/state.go:1570 +0x2ff\ngithub.com/tendermint/tendermint/consensus.(*State).enterCommit.func1()
/home/jake-vega/go/pkg/mod/github.com/vegaprotocol/cometbft@v0.34.28-0.20230322133204-3d8588de736e/consensus/state.go:1505 +0xaa\ngithub.com/tendermint/tendermint/consensus.(*State).enterCommit(0xc002c92380, 0x15a, 0x0)
/home/jake-vega/go/pkg/mod/github.com/vegaprotocol/cometbft@v0.34.28-0.20230322133204-3d8588de736e/consensus/state.go:1543 +0xccf\ngithub.com/tendermint/tendermint/consensus.(*State).addVote(0xc002c92380, 0xc003ce1220, {0xc002c56540, 0x28})
/home/jake-vega/go/pkg/mod/github.com/vegaprotocol/cometbft@v0.34.28-0.20230322133204-3d8588de736e/consensus/state.go:2165 +0x18dc\ngithub.com/tendermint/tendermint/consensus.(*State).tryAddVote(0xc002c92380, 0xc003ce1220, {0xc002c56540?, 0xc0032d4900?})
/home/jake-vega/go/pkg/mod/github.com/vegaprotocol/cometbft@v0.34.28-0.20230322133204-3d8588de736e/consensus/state.go:1963 +0x2c\ngithub.com/tendermint/tendermint/consensus.(*State).handleMsg(0xc002c92380, {{0x3d92c20?, 0xc0041b6290?}, {0xc002c56540?, 0x0?}})
/home/jake-vega/go/pkg/mod/github.com/vegaprotocol/cometbft@v0.34.28-0.20230322133204-3d8588de736e/consensus/state.go:861 +0x170\ngithub.com/tendermint/tendermint/consensus.(*State).receiveRoutine(0xc002c92380, 0x0)
/home/jake-vega/go/pkg/mod/github.com/vegaprotocol/cometbft@v0.34.28-0.20230322133204-3d8588de736e/consensus/state.go:768 +0x3f9\ncreated by github.com/tendermint/tendermint/consensus.(*State).OnStart
/home/jake-vega/go/pkg/mod/github.com/vegaprotocol/cometbft@v0.34.28-0.20230322133204-3d8588de736e/consensus/state.go:379 +0x12d\n"}
```

### Observed behaviour

Consensus failure

### Expected behaviour

Child leaves opening auction. Other two markets are closed/rejected, since only one can define any given parent.

### Steps to reproduce

```bash
On this branch: https://github.com/vegaprotocol/system-tests/tree/jake_successor_market
pytest -k test_first_child_is_only_successor tests/governance/test_markets.py
```

### Software version

e5caa023f5a99dfbf4845e20575b338b4be274b3

### Failing test

tests/governance/test_markets.py:test_first_child_is_only_successor 

### Jenkins run

_No response_

### Configuration used

_No response_

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by EVODelavega**
Closes #8596

### CODE REVIEW

**Review by jeremyletang — DISMISSED** [review]

File: core/execution/engine.go [hunk 1]
```diff
@@ -1013,12 +1011,12 @@ func (e *Engine) OnTick(ctx context.Context, t time.Time) {
...
 	parentStates := e.getParentStates()
 	evts := make([]events.Event, 0, len(e.marketsCpy))
-	rejected := map[int]struct{}{}
+	maxI := len(e.marketsCpy) - 1
 	for i, mkt := range e.marketsCpy {
-		if _, ok := rejected[i]; ok {
-			// successor markets were rejected, because of how golang iterates over slices, this loop
-			// will still iterate over rejected markets, these markets must be skipped
-			continue
+		// when a successor market kicks in, the marketsCpy slice is updated
```
Comment (wwestgarth) [inline core/execution/engine.go:1016]: Can't we just collect the to-reject and reject them after we've finished the loop? Seems generally better than fiddling with the slice while iterating over it.

Comment (EVODelavega) [reply core/execution/engine.go:1016]: Thought of that, but it'd end up being more verbose (we'd have to check every market we iterate over, to see if we should handle the time update or not), and then add an additional loop over that map to reject the markets in question. This way avoids a second loop, results in the same behaviour, and allows for market succession to be handled all in a single block (including cleaning up the checkpoint state etc...).

Comment (wwestgarth) [reply core/execution/engine.go:1016]: More verbose but less risky.

Also won't we already be on-ticking children that should be rejected if they happen to appear before the child that comes out of auction first in the marketsCpy?

Comment (EVODelavega) [reply core/execution/engine.go:1010]: If they appear before the successor, then the `OnTick` is fine. They may be ready to leave opening auction, but in that case, it's the oldest successor that takes priority, so older pending successors must be checked first.

File: core/execution/engine_snapshot_test.go [hunk 2]
```diff
@@ -343,8 +343,9 @@ func TestValidSettledMarketSnapshot(t *testing.T) {
 	marketConfig := getMarketConfig()
 	err := engine.SubmitMarket(ctx, marketConfig, "", time.Now())
 	assert.NoError(t, err)
-	// fake settled market
-	marketConfig.State = types.MarketStateSettled
+	// this does not work. Setting the state to settled will update the state on the market object
+	// unfortunately, it does not set the closed flag, so OnTick will not work
+	// marketConfig.State = types.MarketStateSettled
 	engine.OnTick(ctx, time.Now())
```
Comment (wwestgarth) [inline core/execution/engine_snapshot_test.go:349]: What is `TestValidSettledMarketSnapshot` actually testing then if its not testing the snapshot of a settled market? Maybe the test should be changed to actually make a settled market properly if it wasn't working before.

Comment (EVODelavega) [reply core/execution/engine_snapshot_test.go:349]: Yeah, so there were 2 failing tests, this one is indeed about specifically settled markets. I'll see if I can update the test and properly settle the market in this test

File: core/execution/engine.go [hunk 3]
```diff
@@ -1013,12 +1005,12 @@ func (e *Engine) OnTick(ctx context.Context, t time.Time) {
...
+		// this loop will still iterate over N elements, even if the slice has shrunk
+		if i >= maxI {
```
Comment (wwestgarth) [inline core/execution/engine.go:1012]: The more I think the more that this is not right.

Say have markets [0,1,2,3,4,5] where markets 2,3,4 are all successors of market 1.

We enter loop `i==2` and say market 2 comes out of auction so we reject `3,4` and our slice looks like `[0,1,2,5]`. We set `maxI` to be `len(e.marketsCpy) - 1 == 3`.

We start loop `i=3` and `if i >= maxI` is true so we break, and we've just clean forgotten to process market number 5.

Heres it happening in goplayground: https://go.dev/play/p/J2jZgd8YyhX

I think we should stop playing with fire, and just not play around with the slice while we iterate over it. Collect the markets to reject and remove them later.

Comment (EVODelavega) [reply core/execution/engine.go]: Alright, I've updated the PR. There was still some panic in certain integration tests, apparently the `removeMarket` call in the execution engine did not break out of the loop once the market was found. Not sure if that got accidentally removed, or was a bug that had been there for a while

**Review by wwestgarth — DISMISSED** [review]

**Review by wwestgarth — DISMISSED** [review]

**Review by jeremyletang — APPROVED** [review]

**Pull request merged**